[RyuJIT/armel] Support `double` argument passing

wateret opened the pull request:
Fix #12293

wateret commented on src/jit/lsra.cpp:
```diff
@@ -4148,6 +4161,14 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
             useCandidates  = allRegs(registerType);
         }
 
+#ifdef ARM_SOFTFP
```
This part is kind of pair with lin 421 at laraarmarch.cpp, but I am not sure about this change. @caroleidt this is the part that I was wondering from the our last conversation #12293.

CarolEidt replied:
At this point, the node has been created, so the `gtOtherRegs` values should be correct, right?  I think you could assert `tree->gtOtherRegs[i] == genRegNumFromMask(currCandidates)` below. It would also be good to add a comment that this code, and the code in lsraarmarch.cpp rely on the fact that the registers are in sequential order.

---
In reply to: [122643874](https://github.com/dotnet/coreclr/pull/12360#discussion_r122643874) [](ancestors = 122643874)

wateret replied:
@CarolEidt IMHO `gtOtherRegs`  values are invalidated from Lowering phaseby `node->gtClearReg(comp)`.

CarolEidt replied:
I see. There should at least be a comment about the assumption that the arg registers are in sequential order.

wateret commented on src/jit/lsraarmarch.cpp:
```diff
@@ -403,13 +403,22 @@ void Lowering::TreeNodeInfoInitPutArgReg(
     // To avoid redundant moves, have the argument operand computed in the
     // register in which the argument is passed to the call.
-    node->gtOp.gtOp1->gtLsraInfo.setSrcCandidates(m_lsra, m_lsra->getUseCandidates(node));
+    node->gtOp.gtOp1->gtLsraInfo.setDstCandidates(m_lsra, m_lsra->getUseCandidates(node));
```
To make the other changes(the one with my comment) work, I thought we need this change. Maybe I am still getting something wrong.

CarolEidt replied:
This should be `setSrcCandidates()` because `node` is using its op1 as a source, and setting the requirements on the candidates accordingly.

wateret replied:
Then does it mean that the code "lsra.cpp:4163"(in the loop for pushing produces) should be done in consume loop?

CarolEidt replied:
Yes. This is a somewhat tricky case that I don't think we've handled before - that is, where we are consuming sources from a node that produces multiple registers, and where they have fixed reg requirements. This would need to be handled near line 4013 in your version of lsra.cpp, where it sets `fixedAssignment` assuming that there will only be a single register in that case.

wateret commented:
@dotnet/jit-contrib @CarolEidt Please take a look.

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -1628,6 +1628,9 @@ struct GenTree
+#ifdef _TARGET_ARM_
+            case GT_PUTARG_REG:
+#endif
                 return true;
```
`GT_PUTARG_REG` is defined as `GTK_UNOP`, so I don't see why this is necessary. (i.e., it's not a binary node in the first place)

wateret replied:
I have changed `GT_PUTARG_REG` to be `GenTreeMultiRegOp`(a subclass of `GenTreeOp`) for ARM. That's why it was necessary.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -3803,16 +3806,17 @@ struct GenTreeCmpXchg : public GenTree
-    GenTreeMulLong(var_types type, GenTreePtr op1, GenTreePtr op2) : GenTreeOp(GT_MUL_LONG, type, op1, op2)
+    GenTreeMultiRegOp(var_types type, GenTreePtr op1, GenTreePtr op2)
+        : GenTreeOp(GT_MUL_LONG, type, op1, op2), gtOtherReg(REG_NA)
     {
```
I would assume that, if you changing this from `GenTreeMulLong` to `GenTreeMultiRegOp`, you expect to use this more generally than just for `GT_MUL_LONG`. If that's not the case, then this should be kept as `GenTreeMulLong`. If that is the case, then you shouldn't have an implicit oper on the constructor.

wateret replied:
Oh I have missed that, I'll fix it.

CarolEidt commented:
>     if (IsMultiRegCall())

Not directly related to this PR, but it would seem that this should return true for `GT_COPY` for `FEATURE_MULTIREG_RET`.

---
Refers to: src/jit/gentree.h:5471 in e35a840. [](commit_id = e35a840b793160b0f9047c71338931b564dafbad, deletion_comment = False)

BruceForstall commented:
Can you give an example of the IR (JitDump) before your change? I thought the code already forced all double to long in fgMorphArgs for armel, so I'm not sure why we need special handling for double.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -960,7 +960,16 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+#ifdef _TARGET_ARM_
+            if (comp->opts.compUseSoftFP && type == TYP_DOUBLE)
+            {
+                putArg = new (comp, GT_PUTARG_REG) GenTreeMultiRegOp(type, arg, nullptr);
```
Would this also be required for passing `TYP_LONG`? Also, I think you need to be consistent in using `GenTreeMultiRegOp` for `GT_PUTARG_REG`, or you should use a different oper (e.g. `GT_PUTARG_MULTIREG`), since the JIT (mostly) depends on the oper being a "tag" for the `GenTree` type. That is, a given oper should always be on the same `GenTree` type.

CarolEidt replied:
I realize now that `TYP_LONG` is passed using `GT_FIELD_LIST`, so disregard the first question. However, the other comment still applies.

---
In reply to: [123120550](https://github.com/dotnet/coreclr/pull/12360#discussion_r123120550) [](ancestors = 123120550)

hqueue commented on src/jit/lsra.cpp:
```diff
@@ -3596,7 +3602,14 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
+    // TODO-ARM-Cleanup: we need to change assert condition.
+    // For example, count and dst count of `vmov.d2i` do not match.
+    // The below call is temporally added just to supress warning message.
+    ComputeAvailableSrcCount(tree);
```
Please take a look at [armel CI failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/armel_cross_release_tizen_prtest/3570/console).
You can use it with `DEBUG` defined or in `assert`, because `ComputeAvailableSrcCount()` is defined only when `DEBUG` is defined.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -1175,12 +1184,29 @@ void Lowering::LowerArg(GenTreeCall* call, GenTreePtr* ppArg)
+
+#ifdef _TARGET_ARM_
+            // We need gtOtherRegs[] for the case - copy from a double register to two int registers
+            GenTreePtr intArg = new (comp, GT_COPY) GenTreeCopyOrReload(GT_COPY, intType, arg);
```
You don't need to `ifdef` this. On other targets, although `GenTreeCopyOrReload` is derived from `GenTreeUnOp` it has no additional fields. This is how LSRA generates these nodes.

wateret replied:
It looks like with FEATURE_MULTIREG_RET, there is an additional field. In gentree.h,

```cpp
4923 struct GenTreeCopyOrReload : public GenTreeUnOp
4924 {
4925 #if FEATURE_MULTIREG_RET
4926     // State required to support copy/reload of a multi-reg call node.
4927     // First register is is always given by gtRegNum.
4928     //
4929     regNumberSmall gtOtherRegs[MAX_RET_REG_COUNT - 1];
4930 #endif
```

I wanted to use `gtOtherRegs[0]` as the second register when we copy double register.

CarolEidt replied:
> It looks like with FEATURE_MULTIREG_RET, there is an additional field. In gentree.h

Right. What I meant was that we should always use `GenTreeCopyOrReload` here even without `FEATURE_MULTIREG_RET`, in which case there is no additional cost relative to `GenTreeUnOp`, and it ensures oper/struct consistency for the `GenTree` nodes.

CarolEidt commented on src/jit/lsra.cpp:
```diff
@@ -3596,7 +3602,14 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
+#ifndef ARM_SOFTFP
     assert(((consume == 0) && (produce == 0)) || (ComputeAvailableSrcCount(tree) == consume));
+#else
+    // TODO-ARM-Cleanup: we need to change assert condition.
```
Instead of `TODO-ARM-Cleanup` this should be `TODO-ARM-Bug`

wateret replied:
I'll fix it. Maybe It'd be nicer to change assert condition for ARM.

CarolEidt commented on src/jit/lsra.cpp:
```diff
@@ -10355,6 +10376,23 @@ void TreeNodeInfo::Initialize(LinearScan* lsra, GenTree* node, LsraLocation loca
         dstCandidates = genRegMask(node->gtRegNum);
+#ifdef ARM_SOFTFP
+        regNumber reg2 = REG_NA;
+        if (node->OperGet() == GT_PUTARG_REG)
```
Shouldn't this include `GT_MUL_LONG` as well?

wateret replied:
Yes I think so. However there was no code for `GT_MUL_LONG` here so I'm not sure if I should add that. I'm not sure if there is a case that GT_MUL_LONG node is actually generated. So I left it out.

CarolEidt replied:
If you don't include that node because you don't expect it, then perhaps in the else clause you could assert that it is not a multireg node.

CarolEidt commented on src/jit/lsraarm.cpp:
```diff
@@ -727,6 +727,43 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
+            // TODO-ARM-Cleanup : the logic was copied from general case,
+            // Only the exceptional case was added - when src count should be 2.
+            info->dstCount = tree->IsValue() ? 1 : 0;
+            if (kind & (GTK_CONST | GTK_LEAF))
```
Both of these are known to be `GTK_UNOP`, so this is unnecessary (`GTK_SMPOP` includes `GTK_UNOP`).  I would suggest adding a method to GenTreeMultiRegOp such as `GetRegCount()`, and then you could simply do:
```
info->srcCount = tree->AsGenTreeMultiRegOp()->GetRegCount();
```
You could also assert based on type that `info->srcCount` has the right value.

wateret replied:
@CarolEidt I just tried that but there is no easy way for the tree node to get register count. The reason is kind of similar to https://github.com/dotnet/coreclr/pull/12360/files#r123158463, that is `gtRegNum` and `gtOtherReg` are invalidated from Lowering phase.

CarolEidt commented on src/jit/lsraarmarch.cpp:
```diff
@@ -403,13 +403,22 @@ void Lowering::TreeNodeInfoInitPutArgReg(
+    if (node->TypeGet() == TYP_LONG)
+    {
+        info.srcCount++;
+        argMask |= genRegMask(REG_NEXT(argReg));
```
This would be a good place to assert that `genRegArgNext(argReg == REG_NEXT(argReg)`, and note that the register allocator depends on the arg registers being in sequential order.

CarolEidt commented:
:clock1:

wateret commented:
@BruceForstall JitDump(Main function) of `DblAdd.exe`: [dump_dbladd_armel.txt](https://github.com/dotnet/coreclr/files/1090963/dump_dbladd_armel.txt)
The problem shows at LSRA, not morphing.

wateret commented:
@CarolEidt Added if condition for `GT_MUL_LONG` in lsra.cpp:10379 and removed the unnecessary part(1e977ab).

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -960,7 +960,16 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+#ifdef _TARGET_ARM_
+            if (comp->opts.compUseSoftFP && type == TYP_DOUBLE)
+            {
+                putArg = new (comp, GT_PUTARG_REG) GenTreeMultiRegOp(GT_PUTARG_REG, type, arg, nullptr);
```
In order to reduce `#ifdef`s here, it would be good to add a `Compiler::gtNewPutArgReg()` method.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -1179,8 +1188,14 @@ void Lowering::LowerArg(GenTreeCall* call, GenTreePtr* ppArg)
+#ifdef _TARGET_ARM_
+            // We need gtOtherRegs[] for the case - copy from a double register to two int registers
+            GenTreePtr intArg = new (comp, GT_COPY) GenTreeCopyOrReload(GT_COPY, intType, arg);
+#else  // !_TARGET_ARM_
```
This should not be `#ifdef _TARGET_ARM_`. It should always use `GenTreeCopyOrReload()`

wateret replied:
I see. I'll update it.

wateret commented:
@CarolEidt Please take a look at the last commit. However it does not work. 

It crashes with:

```
buildIntervals second part ========

NEW BLOCK BB01
<RefPosition #0   @0   RefTypeBB BB01 regmask=[]>
at start of tree, map contains: {  }
  N003.                    IL_OFFSET IL offset: 0x0 REG NA
  consume=0 produce=0
     op

Def candidates [allInt], Use candidates [allInt]
at start of tree, map contains: {  }
  N005. t5              =  CNS_DBL   1.0000000000000000 REG NA
  consume=0 produce=1
t5 = op
Interval  1: RefPositions {} physReg:NA Preferences=[allInt]
<RefPosition #1   @5   RefTypeDef <Ivl:1 internal> CNS_DBL BB01 regmask=[allInt]>
Interval  2: RefPositions {} physReg:NA Preferences=[allInt]
<RefPosition #2   @5   RefTypeDef <Ivl:2 internal> CNS_DBL BB01 regmask=[allInt]>

<RefPosition #3   @5   RefTypeUse <Ivl:1 internal> CNS_DBL BB01 regmask=[allInt] last>
<RefPosition #4   @5   RefTypeUse <Ivl:2 internal> CNS_DBL BB01 regmask=[allInt] last>
Def candidates [allDouble], Use candidates [allDouble]
Interval  3: RefPositions {} physReg:NA Preferences=[allDouble]
<RefPosition #5   @6   RefTypeDef <Ivl:3> CNS_DBL BB01 regmask=[allDouble]>
<RefPosition #5   @6   RefTypeDef <Ivl:3> CNS_DBL BB01 regmask=[allDouble]>
at start of tree, map contains: { N005. CNS_DBL -> (6.N005) }
  N007. t7              =  COPY
  consume=2 produce=1
t7 = op
t6 <RefPosition #6   @7   RefTypeFixedReg <Reg:r0 > BB01 regmask=[r0]>
<RefPosition #7   @7   RefTypeUse <Ivl:3> BB01 regmask=[allDouble] last>

Assert failure(PID 21637 [0x00005485], Thread: 21637 [0x5485]): Assertion failed 'iterator != tree->OperandsEnd()' in 'BringUpTest:Main():int'
(IL size 62)

    File: /opt/code/github/hqueue/coreclr/src/jit/lsra.cpp Line: 3947
    Image: /opt/usr/home/owner/dotnet/ttt/corerun
```

Actually I don't understand what is going on in the consume loop(the one that starts with `for (int useIndex = 0; useIndex < consume; useIndex++)`). Could you please explain what is being done from that loop? It looks way more complicated than the produce loop.

CarolEidt commented:
> Could you please explain what is being done from that loop? It looks way more complicated than the produce loop.

It is indeed more complicated than the produce loop. This is because the current node has to match its expected source operands to the previously encountered def nodes in `operandDefs`, and this in turn is made somewhat more complicated by the fact that a single operand may indirectly represent multiple results, if it is contained. So, the outer loop is looping over the number of expected registers used. The first "if" does some bookkeeping on the map of previously seen nodes (`operandDefs`) and then advances to the next operand. The following code goes on to take care of any special handling or constraints on the use.

The assert you are seeing is because `operandDefs` is empty, but it has only found one of two expected source operands. You can see this because the dump indicates that it is handling a `COPY` node that expects to consume 2 sources, but the "at start of tree" message shows a single node:
```
at start of tree, map contains: { N005. CNS_DBL -> (6.N005) }
  N007. t7              =  COPY
  consume=2 produce=1
t7 = op
t6 <RefPosition #6   @7   RefTypeFixedReg <Reg:r0 > BB01 regmask=[r0]>
<RefPosition #7   @7   RefTypeUse <Ivl:3> BB01 regmask=[allDouble] last>
```
So, either the `CNS_DBL` node needs to also produce two registers, or the `COPY` needs to consume just one.

wateret commented:
@CarolEidt With your comment, the last commit looked totally wrong so I removed that commit and added fixes. It also reflects your previous review comments.

wateret commented:
@dotnet-bot Test Tizen armel Cross Release Build Please

CarolEidt commented on src/jit/gentree.cpp:
```diff
@@ -7621,6 +7621,39 @@ GenTree* Compiler::gtNewBlkOpNode(
 }
 
+//------------------------------------------------------------------------
+// gtNewPutArgReg: Creates a new Obj node.
```
This should say "Creates a new PutArgReg node." (or "GT_PUTARG_REG" node, or something like that).

wateret replied:
Sorry I missed that.

CarolEidt commented on src/jit/gentree.cpp:
```diff
@@ -7621,6 +7621,39 @@ GenTree* Compiler::gtNewBlkOpNode(
+//    arg  - The argument node
+//
+// Return Value:
+//    Returns a node representing the struct value at the given address.
```
This comment also needs to be updated.

CarolEidt commented on src/jit/lsra.cpp:
```diff
@@ -3596,7 +3602,13 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
+#ifndef ARM_SOFTFP
     assert(((consume == 0) && (produce == 0)) || (ComputeAvailableSrcCount(tree) == consume));
+#else
+    assert(((consume == 0) && (produce == 0)) || ((consume == 1) && (produce == 2) && tree->OperGet() == GT_COPY &&
```
This is not correct. It should always be the case that either the node consumes and produces no registers, or the latter condition holds. If this doesn't hold, then you have the wrong count on one of the nodes. It looks like the code below sets the `srcCount` to 1 and the `dstCount` to 2 for this case of `GT_COPY`, so why is this special case needed?

wateret replied:
It is only for last drafts, now we don't need that. Thanks!

CarolEidt commented on src/jit/lsraarm.cpp:
```diff
@@ -727,6 +727,31 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
+
+        case GT_PUTARG_REG:
+#ifdef ARM_SOFTFP
+            if (tree->TypeGet() == TYP_LONG) // Actually was `double` type
```
How is an actual long type parameter passed? Wouldn't that also be `TYP_LONG`? Or would it have been decomposed into to `TYP_INT` parameters?

wateret replied:
As far as I can see, `TYP_LONG`s must have been decomposed like you said.

CarolEidt replied:
It might be worth an extra comment

CarolEidt replied:
(e.g. "This case currently only occurs for double types that are passed as TYP_LONG; actual long types would have been decomposed by now."

CarolEidt commented on src/jit/lsraarmarch.cpp:
```diff
@@ -403,7 +403,17 @@ void Lowering::TreeNodeInfoInitPutArgReg(
 
     // Set the register requirements for the node.
-    const regMaskTP argMask = genRegMask(argReg);
+    regMaskTP argMask = genRegMask(argReg);
```
In looking through the code, I can't see where this method (`Lowering::TreeNodeInfoInitPutArgReg()`) is called. It appears that for arm64 it is handled with the default case, and for arm it is handled directly in `Lowering::TreeNodeInfoInit()`.

wateret replied:
It is called from `Lowering::TreeNodeInfoInitCall(GenTreeCall* call)`. In `Lowering::TreeNodeInfoInitPutArgReg()`, not only it sets the PutArgReg node's candidates but also modifies the Call node's srcCount(`TreeNodeInfo& info` is the call node's info).

CarolEidt replied:
I see. We may need to modify that in future, because once we are doing the containment analysis in `Lowering`, and then doing `TreeNodeInfoInit` on the fly, we need to be able to set all the counts and candidates in a forward walk (without modifying a node's info from its parent). But this works for now.

CarolEidt commented:
@wateret - This looks good; thanks for the additional comments. It seems that there are now conflicts that need to be resolved and then we can merge this.

wateret commented:
@CarolEidt I just resolved conflicts and squashed commits to clean up the commit message. Thank you for your review :D

wateret commented:
@dotnet-bot Test Tizen armel Cross Debug Build please
